### PR TITLE
Selection tier

### LINKED ISSUE #16555: Switch Engineer/Harvester Selection Priorities (Quality of Life)
Engineer Selection Priority: 5
Harvester Selection Priority: 7

Basically, change it so engineers are selected _before_ harvesters. This is in response to a rare issue I've had, and recently learned others do too, that in a panic someone sends their harvesters instead of their engineer.

I'm not 100% sure the repercussions of this (what is priority 6?), but the goal is to change it so, when harvesters/engineers are in the same selection box, engineers are selected, not harvesters.

Harvesters are very easy to select, while engineers are not, so I don't see any issue with making this change.

### PR DESCRIPTION
**Pull request by asfreitas**
I made some changes to the selection tier based on suggestions made and I also made a new branch since I had forgotten to the first time around. Closes #16555

### CODE REVIEW

**Comment by netnazgul** [comment]
Can you re-add the simplified selection tiers table in the original post? So it's easier to check.

Also wrt the PR, you'd likely need to squash all the commits into a single one here as they are functionally indivisible.

File: mods/ra/rules/defaults.yaml
Comment (netnazgul) [inline mods/ra/rules/defaults.yaml]: These empty lines (a lot of them also removed in other places) should be reverted as they were for YAML readability (they split the code blocks).

File: mods/ra/rules/defaults.yaml
Comment (netnazgul) [inline mods/ra/rules/defaults.yaml]: A spacer empty line missing here.

**Review by netnazgul — CHANGES_REQUESTED** [review]

**Comment by netnazgul** [comment]
Also, should we streamline TD/D2K selection tiers as well in this PR?

**Comment by abcdefg30** [comment]
Sounds like a good idea.

File: mods/ra/rules/vehicles.yaml
Comment (netnazgul) [inline mods/ra/rules/vehicles.yaml]: Can remove this line as `Selectable:` is only mentioned for `Priority:` here, and we delete that.

File: mods/ra/rules/infantry.yaml
Comment (netnazgul) [inline mods/ra/rules/infantry.yaml]: ^

File: mods/ra/rules/structures.yaml
Comment (netnazgul) [inline mods/ra/rules/structures.yaml]: ^

File: mods/ra/rules/vehicles.yaml
Comment (netnazgul) [inline mods/ra/rules/vehicles.yaml]: ^

**Review by netnazgul — CHANGES_REQUESTED** [review]

**Comment by netnazgul** [comment]
So if I didn't miss anything, this is the list for TD:

- CombatUnit: `^Soldier` and `^Vehicle`
- SupportUnit: E6, MCV
- EcoUnit: HARV, TRUCK 
- CombatBuilding: GUN, SAM, OBLI, GTWR, ATWR (Can be attached to `^Defence`)
- Building: everything else (can be attached to `^BaseBuilding`); there is also `^Wall`, but I guess those are not selectable, right?

**Comment by netnazgul** [comment]
@MustaphaTR requested to fill the list for D2K

**Comment by asfreitas** [comment]
Yeah, this is great. I'll make the changes to TD. But I've never played D2K so I'll definitely need some help there.

**Comment by dragunoff** [comment]
@asfreitas For D2k:

Tier | Actors
:-- | :--
Support Units | `engineer`, `mcv`
Economic Units | `harvester`
Combat Buildings | `^Defense`
Buildings | `^Building`

Note: you should remove the field `Priority` from `medium_gun_turret` and `large_gun_turret`.

And @abcdefg30 suggested that `thumper` be also added to support units. And I'm wondering about `saboteur` but I'm not so familiar with the game so someone with better understanding should weigh in.

**Comment by asfreitas** [comment]
@dragunoff 

I'm getting the following error when trying to run D2K:
Internal mods:
        all: All mods ({DEV_VERSION})
        cnc: Tiberian Dawn ({DEV_VERSION})
        d2k: Dune 2000 ({DEV_VERSION})
        modcontent: Mod Content Manager ({DEV_VERSION})
        ra: Red Alert ({DEV_VERSION})
        ts: Tiberian Sun ({DEV_VERSION})
External mods:
        cnc-{DEV_VERSION}: Tiberian Dawn ({DEV_VERSION})
        d2k-{DEV_VERSION}: Dune 2000 ({DEV_VERSION})
        ra-{DEV_VERSION}: Red Alert ({DEV_VERSION})
Loading mod: d2k
Exception of type `System.IO.FileNotFoundException`: Assembly `d2k|OpenRA.Mods.D2k.dll` not found.
   at OpenRA.ObjectCreator..ctor(Manifest manifest, InstalledMods mods)
   at OpenRA.ModData..ctor(Manifest mod, InstalledMods mods, Boolean useLoadScreen)
   at OpenRA.Game.InitializeMod(String mod, Arguments args)
   at OpenRA.Game.Initialize(Arguments args)
   at OpenRA.Game.InitializeAndRun(String[] args)
   at OpenRA.Program.Main(String[] args)

Have you ever seen anything like this before?

**Comment by abcdefg30** [comment]
Hm, that looks like the dll was not compiled. Can you recompile and look if the `OpenRA.Mods.D2k.dll` file gets created (inside the d2k folder)?

**Comment by asfreitas** [comment]
> Hm, that looks like the dll was not compiled. Can you recompile and look if the `OpenRA.Mods.D2k.dll` file gets created (inside the d2k folder)?

@abcdefg30 Yeah it looks like it's being created and it's in OpenRA\OpenRA.Mods.D2k\obj\x86\Debug. After recompile it's still crashing and saying it cannot find the file.

**Comment by netnazgul** [comment]
@asfreitas are you trying to run a debug build from VS? Just do `make all` instead and then run the game via OpenRA.Game.exe as usual if you don't explicitly need the actual debug mode.

**Comment by asfreitas** [comment]
@netnazgul Thanks for the heads up. Using "make all" fixed the problem. I finished the changes and it should be good to go now.

**Comment by abcdefg30** [comment]
Unfortunately you will have to rebase this PR (and squash the commits), otherwise we can't merge the PR (see the "This branch has conflicts that must be resolved" warning below). If you are new to rebasing, https://help.github.com/en/articles/about-git-rebase might be a good starting point. (If that proves too difficult you could reset this branch and reapply your changes.) We can also help you on IRC if you want.

**Comment by asfreitas** [comment]
@abcdefg30 Thanks for your help! I saw that we needed to rebase but I was a bit unfamiliar with what that meant with regard for merge conflicts. Makes sense now. If I need to do anything else please let me know!

**Comment by abcdefg30** [comment]
> If I need to do anything else please let me know!

It would be good if you could squash the commits (http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html if you didn't see it before), we try to avoid merging "fixup" commits. (I.e. it would be good if you could squash this into one commit, or 3 commits with one for ra, one for td and one for d2k if you wish.)

**Comment by abcdefg30** [comment]
Oh and I just saw that some of @netnazgul's remarks are not resolved. You probably want to fix them too. (Bonus points if you change `Inherits@selection:` to `Inherits@SELECTION:`, but I don't think that is a blocker.)

**Comment by dragunoff** [comment]
> 
> 
> (Bonus points if you change `Inherits@selection:` to `Inherits@SELECTION:`, but I don't think that is a blocker.)
>
Seems like we really don't have a convention with regards to case in these blocks (see https://github.com/OpenRA/OpenRA/pull/16612#discussion_r289202642) :-) And my PR for selection modes was merged with `@SELECTION_MODE`

**Comment by abcdefg30** [comment]
Ah yes, I thought we upper case everything. Can leave it then I guess.

**Review by abcdefg30 — APPROVED** [review]
I took the freedom to rebase your merge commit away (and added one missing new line). I hope you don't mind. This looks good to go to me now. :+1:

**Comment by dragunoff** [comment]
In TD:
  * Please add to the Supply Truck (`TRUK`) the economic unit prioriry

In D2k:
  * Please add `Inherits@SELECTION_MODE: ^LowPrioritySelectionMode` to the engineer to mactch the seleciton mode to the priority

Looks good in RA. I think we should cover TS as well.

**Comment by dragunoff** [comment]
For TS:

Tier | Actors
:-- | :--
Combat Units | `^Infantry`, `^Vehicle`, `^Aircraft`, `^Visceroid`, `^Train`
Support Units | `ENGINEER`, `MCV`
Economic Units | `HARV`
Combat Buildings | `^Defense`
Buildings | `^BasicBuilding`

**Comment by dragunoff** [comment]
:thinking: I took a deeper look and I noticed that `^SelectableCombatUnit` is not used consistently. For exapmle it is defined in RA but not inherited anywhere. In TD it's not inherited by all selectable actors.

Also there is a overlap between this and the recently merged selection modes. Perhaps we should consider merging these in a single declaration? For example:
```yaml
^SelectableCombatUnit:
	Selectable:
		Priority: 10
		PriorityModifiers: Ctrl

^SelectableSupportUnit:
	Selectable:
		Priority: 8
		PriorityModifiers: Ctrl, Alt

^SelectableEconomicUnit:
	Selectable:
		Priority: 6
		PriorityModifiers: Ctrl, Alt

^SelectableCombatBuilding:
	Selectable:
		Priority: 4

^SelectableBuilding:
	Selectable:
		Priority: 2
```

**Comment by asfreitas** [comment]
Thanks for the help and feedback! I'm sorry I thought that I could help more after exams finished but then I got overwhelmed at work and rebasing was new to me--I wasn't sure why I wasn't "seeing" my last commit and so I wasn't sure about squashing the commit. I'm sure someone else can take over now but I'd love to help out in the future when I have more time. Good luck and all the best.

**Comment by dragunoff** [comment]
I can take over this.

**Comment by abcdefg30** [comment]
Superseded by #16760 now.

**Pull request closed without merging**
